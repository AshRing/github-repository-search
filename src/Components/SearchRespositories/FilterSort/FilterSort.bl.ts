export const sortInputs = {
	bestMatch: "Best Match",
	stars: "Stars",
};

export const availableSorts = Object.keys(sortInputs).map((key: string) => sortInputs[key]);

export const sortValues = {
	[sortInputs.bestMatch]: undefined,
	[sortInputs.stars]: "stars",
};

export const filters = {
	language: "language",
};

export const availableFilters = Object.keys(filters).map((key: string) => filters[key]);

export const languageFilterOptions = [
	"Assembly",
	"C",
	"C#",
	"C++",
	"Dart",
	"HTML",
	"Haskell",
	"Go",
	"Java",
	"JavaScript",
	"Python",
	"Ruby",
	"Rust",
	"Shell",
	"TypeScript",
];
