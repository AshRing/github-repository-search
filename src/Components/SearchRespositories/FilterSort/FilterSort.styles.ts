import styled from "styled-components";

export const SearchInputContainer = styled.div`
	position: relative;
	width: 100%;
	margin-bottom: 1rem;

	& > .fa-search {
		position: absolute;
		top: 50%;
		left: 1rem;
		transform: translate(-50%, -50%);
		color: ${({ theme }) => theme.colors.secondary};
	}
`;

export const SearchInput = styled.input`
	width: 100%;
	padding: 0 1rem 0 2rem;
	height: 2.5rem;
	border-radius: 0.25rem;
	border: 1px solid ${({ theme }) => theme.colors.secondary};
	outline: none;
	transition: 0.2s ease border-color;

	&:focus {
		border-color: ${({ theme }) => theme.colors.primary};
	}
`;

export const FilterSortContainer = styled.div`
	display: flex;
	column-gap: 1rem;

	& > div {
		width: 50%;
	}
`;
