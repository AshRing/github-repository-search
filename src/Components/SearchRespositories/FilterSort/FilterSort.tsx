import * as React from "react";

import {
	FilterSortContainer,
	FilterSortGroup,
	SearchInput,
	SearchInputContainer,
} from "./FilterSort.styles";
import { availableSorts, filters, languageFilterOptions } from ".";
import { faSearch, faTimes } from "@fortawesome/free-solid-svg-icons";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { IFilterSortOption } from "src/_types";
import { SelectInput } from "../../../Components/Shared";

interface Props {
	addFilter: (filterToAdd: IFilterSortOption) => void;
	changeSearchTerm: (value: string) => void;
	changeSort: (sortName: string) => void;
	filterBy: IFilterSortOption[];
	searchRepositories: () => void;
	searchTerm: string;
	sortBy: string;
}

export const FilterSort = ({
	addFilter,
	changeSearchTerm,
	changeSort,
	filterBy,
	searchRepositories,
	searchTerm,
	sortBy,
}: Props) => {
	const searchInputRef: React.RefObject<HTMLInputElement> = React.useRef();

	React.useEffect(() => {
		searchInputRef.current.focus();
	}, []);

	return (
		<FilterSortContainer
			onSubmit={(e) => {
				e.preventDefault();
				searchRepositories();
			}}
		>
			<SearchInputContainer>
				<FontAwesomeIcon icon={faSearch} />
				<SearchInput
					ref={searchInputRef}
					type="text"
					placeholder="Search Repositories"
					value={searchTerm}
					onChange={(e: React.ChangeEvent) => changeSearchTerm(e.target.value)}
				/>
				{searchTerm && (
					<button type="button" onClick={() => changeSearchTerm("")}>
						<FontAwesomeIcon icon={faTimes} />
					</button>
				)}
			</SearchInputContainer>
			<FilterSortGroup>
				<div>
					<SelectInput
						multiselect
						label="Language"
						selected={
							filterBy.find(
								(filter: IFilterSortOption) => filter.name === filters.language,
							)?.values || []
						}
						options={languageFilterOptions}
						handleChange={(selected: string[]) =>
							addFilter({ name: filters.language, values: selected })
						}
					/>
				</div>
				<div>
					<SelectInput
						label="Sort"
						selected={[sortBy]}
						options={availableSorts}
						handleChange={(selected: string[]) => changeSort(selected[0])}
					/>
				</div>
			</FilterSortGroup>
		</FilterSortContainer>
	);
};
