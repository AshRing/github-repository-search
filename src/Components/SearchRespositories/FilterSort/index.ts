export * from "./FilterSort";
export * from "./FilterSort.bl";
