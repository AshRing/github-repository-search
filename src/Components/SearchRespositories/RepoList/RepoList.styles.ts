import styled from "styled-components";

export const RepoListContainer = styled.div`
	padding-top: 1rem;
	border-top: 1px solid ${({ theme }) => theme.colors.secondary};
`;
