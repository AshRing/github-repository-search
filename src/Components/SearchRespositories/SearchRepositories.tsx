import * as React from "react";

import { CHANGE_SEARCH_TERM, searchReposInitialState, searchReposReducer } from "./reducer";
import {
	FilterSortContainer,
	SearchInput,
	SearchInputContainer,
	SearchRepositoriesContainer,
} from "./SearchRepositories.styles";

import { FilterInput } from "./FilterInput";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { SortInput } from "./SortInput";
import { faSearch } from "@fortawesome/free-solid-svg-icons";

export const SearchRepositories = () => {
	const [searchReposState, searchReposDispatch] = React.useReducer(
		searchReposReducer,
		searchReposInitialState,
	);

	return (
		<SearchRepositoriesContainer>
			<h1>GitHub Repository Search</h1>
			<SearchInputContainer>
				<FontAwesomeIcon icon={faSearch} />
				<SearchInput
					type="text"
					placeholder="Search Repositories"
					value={searchReposState.searchTerm}
					onChange={(e: React.ChangeEvent) =>
						searchReposDispatch({
							type: CHANGE_SEARCH_TERM,
							searchTerm: e.target.value,
						})
					}
				/>
			</SearchInputContainer>
			<FilterSortContainer>
				<FilterInput
					filterBy={searchReposState.filterBy}
					onFilterChange={(selectedFilters: string[]) => console.log(selectedFilters)}
				/>
				<SortInput
					sortBy={searchReposState.sortBy}
					onSortChange={(selectedSort: string) => console.log(selectedSort)}
				/>
			</FilterSortContainer>
		</SearchRepositoriesContainer>
	);
};
