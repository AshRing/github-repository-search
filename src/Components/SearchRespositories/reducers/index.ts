export * from "./filterSortReducer";
export * from "./repoReducer";
