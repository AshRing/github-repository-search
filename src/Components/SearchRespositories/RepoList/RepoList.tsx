import * as React from "react";

import { RepoListContainer, StatusText } from "./RepoList.styles";

import { IRepository } from "../../../_types";
import { Link } from "react-router-dom";
import { Pagination } from "./Pagination";
import { Repo } from "./Repo";

interface Props {
	changePage: (pageNum: number) => void;
	pageNum: number;
	repos: IRepository[];
	reposLoading: boolean;
	totalPages: number;
}

export const RepoList = ({ changePage, pageNum, repos, reposLoading, totalPages }: Props) => {
	const renderStatusText = () => {
		if (reposLoading) {
			return <StatusText>Loading...</StatusText>;
		}

		if (!repos) {
			return <StatusText>Please enter a search term</StatusText>;
		}
	};

	return (
		<RepoListContainer>
			{renderStatusText()}
			{!reposLoading && (
				<>
					{repos?.map((repo: IRepository) => (
						<Link to={`/${repo.owner.login}/${repo.name}`} key={repo.id}>
							<Repo repoInfo={repo} />
						</Link>
					))}
					{repos?.length > 0 && (
						<Pagination
							changePage={changePage}
							currentPage={pageNum}
							totalPages={totalPages}
						/>
					)}
				</>
			)}
		</RepoListContainer>
	);
};
