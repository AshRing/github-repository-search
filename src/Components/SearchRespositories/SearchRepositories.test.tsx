import * as React from "react";
import * as apiCalls from "../../_api/getRepositories";
import * as routeData from "react-router-dom";
import * as searchRepoBl from "./SearchRepositories.bl";

import { FilterSort, filters, sortInputs } from "./FilterSort";
import { TestWrapper, mockRepo1 } from "../../_mocks";
import { filterSortInitialState, repoReducerInitialState } from "./reducers";
import { mount, shallow } from "enzyme";

import { IFilterSortOption } from "../../_types";
import { RepoList } from "./RepoList";
import { SearchRepositories } from ".";
import { act } from "react-dom/test-utils";

const waitForComponentToPaint = async (wrapper) => {
	await act(async () => {
		await new Promise((resolve) => setTimeout(resolve));
		wrapper.update();
	});
};

const mockIntersectionObserver = jest.fn();
mockIntersectionObserver.mockReturnValue({
	observe: () => null,
	unobserve: () => null,
	disconnect: () => null,
});
window.IntersectionObserver = mockIntersectionObserver;

const locationSearch = "";
jest.mock("react-router-dom", () => ({
	__esModule: true,
	...(jest.requireActual("react-router-dom") as {}),
	useHistory: () => ({
		replace: jest.fn(),
	}),
	useLocation: () => ({
		state: {
			search: locationSearch,
		},
	}),
}));

const getRepositoriesSpy = jest
	.spyOn(apiCalls, "getRepositories")
	.mockResolvedValue({ total_count: 1, items: [mockRepo1] });

afterEach(() => {
	jest.spyOn(routeData, "useLocation").mockReturnValue({
		state: {},
		pathname: "/",
		hash: "",
		search: locationSearch,
	});
});

test("SearchRepositories matches snapshot", () => {
	const wrapper = shallow(<SearchRepositories />);
	expect(wrapper).toMatchSnapshot();
});

describe("query params", () => {
	test("applies query params - search term", async () => {
		const searchTerm = "test";
		const filter: IFilterSortOption = { name: filters.language, values: ["HTML"] };
		const sortBy = sortInputs.stars;
		jest.spyOn(routeData, "useLocation").mockReturnValue({
			state: {},
			pathname: "/",
			hash: "",
			search: `?searchTerm=${searchTerm}&${filter.name}=${filter.values[0]}&sortBy=${sortBy}`,
		});

		const wrapper = mount(
			<TestWrapper>
				<SearchRepositories />
			</TestWrapper>,
		);

		await waitForComponentToPaint(wrapper);

		expect(getRepositoriesSpy).toHaveBeenCalled();
		expect(wrapper.find(FilterSort).prop("filterBy")).toEqual([filter]);
		expect(wrapper.find(FilterSort).prop("sortBy")).toEqual(sortBy);
		expect(wrapper.find(FilterSort).prop("searchTerm")).toEqual(searchTerm);
	});

	test("applies query params - no search term", async () => {
		const filter: IFilterSortOption = { name: filters.language, values: ["HTML"] };
		const sortBy = sortInputs.stars;
		jest.spyOn(routeData, "useLocation").mockReturnValue({
			state: {},
			pathname: "/",
			hash: "",
			search: `?${filter.name}=${filter.values[0]}&sortBy=${sortBy}`,
		});
		const wrapper = mount(
			<TestWrapper>
				<SearchRepositories />
			</TestWrapper>,
		);

		await waitForComponentToPaint(wrapper);

		expect(getRepositoriesSpy).toHaveBeenCalled();
		expect(wrapper.find(FilterSort).prop("filterBy")).toEqual([filter]);
		expect(wrapper.find(FilterSort).prop("sortBy")).toEqual(sortBy);
		expect(wrapper.find(FilterSort).prop("searchTerm")).toEqual(
			filterSortInitialState.searchTerm,
		);
	});
});

describe("search term", () => {
	test("changes search term", async () => {
		const searchTerm = "tetris";
		const updateQuerySpy = jest.spyOn(searchRepoBl, "updateQueryParam");
		const wrapper = mount(
			<TestWrapper>
				<SearchRepositories />
			</TestWrapper>,
		);

		await waitForComponentToPaint(wrapper);

		wrapper.find(FilterSort).prop("changeSearchTerm")(searchTerm);
		wrapper.update();
		expect(updateQuerySpy).toHaveBeenCalledWith(
			locationSearch,
			"searchTerm",
			expect.any(Function),
			[searchTerm],
		);
		expect(wrapper.find(FilterSort).prop("searchTerm")).toEqual(searchTerm);
	});

	test("resets repos if search term becomes empty", async () => {
		const searchTerm = "tetris";
		const updateQuerySpy = jest.spyOn(searchRepoBl, "updateQueryParam");
		const wrapper = mount(
			<TestWrapper>
				<SearchRepositories />
			</TestWrapper>,
		);

		await waitForComponentToPaint(wrapper);

		wrapper.find(RepoList).prop("changePage")(2);
		wrapper.find(FilterSort).prop("changeSearchTerm")(searchTerm);
		wrapper.update();
		expect(updateQuerySpy).toHaveBeenCalledWith(
			locationSearch,
			"searchTerm",
			expect.any(Function),
			[searchTerm],
		);
		expect(wrapper.find(FilterSort).prop("searchTerm")).toEqual(searchTerm);

		wrapper.find(FilterSort).prop("changeSearchTerm")("");
		wrapper.update();

		await waitForComponentToPaint(wrapper);
		expect(updateQuerySpy).toHaveBeenCalledWith(
			locationSearch,
			"searchTerm",
			expect.any(Function),
			[],
		);
		expect(wrapper.find(FilterSort).prop("searchTerm")).toEqual("");
		expect(wrapper.find(RepoList).prop("pageNum")).toEqual(repoReducerInitialState.pageNum);
	});
});

test("changes sort", async () => {
	const sortBy = sortInputs.stars;
	const updateQuerySpy = jest.spyOn(searchRepoBl, "updateQueryParam");
	const wrapper = mount(
		<TestWrapper>
			<SearchRepositories />
		</TestWrapper>,
	);

	await waitForComponentToPaint(wrapper);

	wrapper.find(FilterSort).prop("changeSort")(sortBy);
	wrapper.update();
	expect(updateQuerySpy).toHaveBeenCalledWith(locationSearch, "sortBy", expect.any(Function), [
		sortBy,
	]);
	expect(wrapper.find(FilterSort).prop("sortBy")).toEqual(sortBy);
});

test("changes filter", async () => {
	const filter: IFilterSortOption = { name: filters.language, values: ["HTML"] };
	const updateQuerySpy = jest.spyOn(searchRepoBl, "updateQueryParam");
	const wrapper = mount(
		<TestWrapper>
			<SearchRepositories />
		</TestWrapper>,
	);
	await waitForComponentToPaint(wrapper);

	wrapper.find(FilterSort).prop("addFilter")(filter);
	wrapper.update();
	expect(updateQuerySpy).toHaveBeenCalledWith(
		locationSearch,
		filter.name,
		expect.any(Function),
		filter.values,
	);
	expect(wrapper.find(FilterSort).prop("filterBy")).toEqual([filter]);
});

test("search repositories prop makes api request to search repositories", async () => {
	const wrapper = mount(
		<TestWrapper>
			<SearchRepositories />
		</TestWrapper>,
	);
	await waitForComponentToPaint(wrapper);

	wrapper.find(FilterSort).prop("searchRepositories")();
	expect(getRepositoriesSpy).toHaveBeenCalled();
});

test("if search term is present, changing pageNum will initiate repository api request", async () => {
	const searchTerm = "test";
	const wrapper = mount(
		<TestWrapper>
			<SearchRepositories />
		</TestWrapper>,
	);
	await waitForComponentToPaint(wrapper);

	wrapper.find(FilterSort).prop("changeSearchTerm")(searchTerm);
	wrapper.find(RepoList).prop("changePage")(2);
	expect(getRepositoriesSpy).toHaveBeenCalled();
});

test("changes page", async () => {
	const pageNum = 2;
	const updateQuerySpy = jest.spyOn(searchRepoBl, "updateQueryParam");
	const wrapper = mount(
		<TestWrapper>
			<SearchRepositories />
		</TestWrapper>,
	);
	await waitForComponentToPaint(wrapper);

	wrapper.find(RepoList).prop("changePage")(pageNum);
	wrapper.update();
	expect(updateQuerySpy).toHaveBeenCalledWith(locationSearch, "pageNum", expect.any(Function), [
		pageNum.toString(),
	]);
	expect(wrapper.find(RepoList).prop("pageNum")).toEqual(pageNum);
});
