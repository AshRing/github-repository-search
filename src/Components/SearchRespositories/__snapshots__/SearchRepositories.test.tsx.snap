// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchRepositories matches snapshot 1`] = `ShallowWrapper {}`;
